File: kafka-0.7.0/kafka-0.7.0-core-java/src/main/java/kafka/utils/KafkaScheduler.java
package kafka.utils;

import org.apache.log4j.Logger;

import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

public class KafkaScheduler {
    private Logger logger = Logger.getLogger(getClass());
    
    private AtomicLong threadId = new AtomicLong(0);
    
    private int numThreads;
    private String baseThreadName;
    private Boolean isDaemon;
    
    private ScheduledThreadPoolExecutor executor;
    
    
    public KafkaScheduler(int numThreads, String baseThreadName, Boolean isDaemon) {
        this.numThreads = numThreads;
        this.baseThreadName = baseThreadName;
        this.isDaemon = isDaemon;
        
        // 初始化线程池
        this.executor = new ScheduledThreadPoolExecutor(numThreads, new ThreadFactory() {
            public Thread newThread(Runnable runnable) {
                Thread t = new Thread(runnable, baseThreadName + threadId.getAndIncrement());
                t.setDaemon(isDaemon);
                return t;
            }
        });
        executor.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
        executor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
    }
    
    
    public void scheduleWithRate(UnitFunction fun, Long delayMs, Long periodMs) {
        executor.scheduleAtFixedRate(Utils.loggedRunnable(fun), delayMs, periodMs, TimeUnit.MILLISECONDS);
        
    }
    
    public void shutdownNow() {
        executor.shutdownNow();
        logger.info("force shutdown scheduler " + baseThreadName);
    }
    
    public void shutdown() {
        executor.shutdown();
        logger.info("shutdown scheduler " + baseThreadName);
    }
    
}

File: kafka-0.7.0/kafka-0.7.0-core/src/main/scala/kafka/utils/KafkaScheduler.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.utils

import java.util.concurrent._
import java.util.concurrent.atomic._
import kafka.utils._
import org.apache.log4j.Logger

/**
 * A scheduler for running jobs in the background
 * TODO: ScheduledThreadPoolExecutor notriously swallows exceptions
 */
class KafkaScheduler(val numThreads: Int, val baseThreadName: String, isDaemon: Boolean) {
  private val logger = Logger.getLogger(getClass())

  private val threadId = new AtomicLong(0)
  // 创建线程池
  private val executor = new ScheduledThreadPoolExecutor(numThreads, new ThreadFactory() {
    def newThread(runnable: Runnable): Thread = {
      val t = new Thread(runnable, baseThreadName + threadId.getAndIncrement)
      t.setDaemon(isDaemon)
      t
    }
  })

  executor.setContinueExistingPeriodicTasksAfterShutdownPolicy(false)
  executor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false)

  def scheduleWithRate(fun: () => Unit, delayMs: Long, periodMs: Long) =
    executor.scheduleAtFixedRate(Utils.loggedRunnable(fun), delayMs, periodMs, TimeUnit.MILLISECONDS)

  def shutdownNow() {
    executor.shutdownNow()
    logger.info("force shutdown scheduler " + baseThreadName)
  }

  def shutdown() {
    executor.shutdown()
    logger.info("shutdown scheduler " + baseThreadName)
  }
}
